cloud/store/migration.go

```diff
@@ -0,0 +1,178 @@
+		log.Info("current RAFT based schema has data and no migration has been recovered, skipping migration")
+		return nil
+	}
+
+	log.Info(fmt.Sprintf("Migrating %d classes from non-raft to raft based representation", len(schema.Classes)))
+	// Before starting the migration, write the progress file at least once to ensure that we will still recover the
+	// migration if we crash between the first import into RAFT and the write of that update to disk.
+	err = os.WriteFile(migrationProgressFile, []byte{}, raftMigrationFilePermissions)
+	if err != nil {
+		return fmt.Errorf("could not write migration progress to %s: %w", migrationProgressFile, err)
+	}
+	// Apply the change from non-raft to raft based schema, applying change with "schema only" to true to avoid
+	// re-updating the underlying representation
+	for _, class := range schema.Classes {
+		if class == nil {
+			log.Warn("found nil class in non-raft based schema, continuing migration")
+			continue
+		}
+
+		// Skip class if it has been already migrated
+		if _, ok := migrationProgress.MigratedClasses[class.Class]; ok {
+			log.Info(fmt.Sprintf("skipping migrating class %s as it is marked as already migrated", class.Class))
+			continue
+		}
+
+		log.Debug(fmt.Sprintf("migrating class %s to raft based schema", class.Class))
+
+		// Migrate the class from schemaRepo to RAFT
+		marshalledSubCmd, err := json.Marshal(&cmd.AddClassRequest{
+			Class: class,
+			State: shardingState[class.Class],
+		})
+		if err != nil {
+			return fmt.Errorf("could not marshall add class request to bytes: %w", err)
+		}
+		err = st.Execute(&cmd.ApplyRequest{
+			Type:       cmd.ApplyRequest_TYPE_ADD_CLASS,
+			Class:      class.Class,
+			SubCommand: marshalledSubCmd,
+		})
+		// That check if necessary if we are recovering from a migration where the migration progress file got corrupted
+		// and we are re-importing class that already exists in the raft schema
+		if err != nil && errors.Is(err, errClassExists) {
+			log.Info(fmt.Sprintf("class %s already exists in schema: skipping migration", class.Class))
+		} else if err != nil {
+			return fmt.Errorf("could not add class %s to raft based schema during migration: %w", class.Class, err)
+		}
+
+		// Update migrationProgress with the newly migrated class and write to disk
+		migrationProgress.MigratedClasses[class.Class] = true
+		jsonData, err := json.Marshal(migrationProgress)
+		if err != nil {
+			return fmt.Errorf("could not marshall migration progress: %w", err)
+		}
+		err = os.WriteFile(migrationProgressFile, jsonData, raftMigrationFilePermissions)
+		if err != nil {
+			return fmt.Errorf("could not write migration progress to %s: %w", migrationProgressFile, err)
+		}
+		log.Debug(fmt.Sprintf("successfully migrated class %s to raft based schema", class.Class))
+	}
+
+	// Migration complete successfully, ensure we remove the migration file so that we don't try any further migration
+	err = os.Remove(migrationProgressFile)
```

Comment:
**suggestion (if-minor):** maybe we rename it and do use it check after the migration done, if the migration left any class to make sure before and after are equal 🤔 ?

Reply:
That could be a good idea to have a "check" or "verify" step after the migration. But I'm not sure how it would look like/what it would do. Just read the schema again and check against the current state ?

Reply:
yes, like read the schema in memory and compare it with the generated file output byte to byte